src/arkode/arkode_impl.h

```diff
@@ -308,6 +345,40 @@ struct ARKodeMemRec
   ARKTimestepInitFn step_init;
   ARKTimestepFullRHSFn step_fullrhs;
   ARKTimestepStepFn step;
+  ARKTimetepSetUserDataFn step_setuserdata;
+  ARKTimestepPrintAllStats step_printallstats;
+  ARKTimestepWriteParameters step_writeparameters;
+  ARKTimestepResize step_resize;
+  ARKTimestepReset step_reset;
+  ARKTimestepFree step_free;
+  ARKTimestepPrintMem step_printmem;
+  ARKTimestepSetDefaults step_setdefaults;
+  ARKTimestepComputeState step_computestate;
+  ARKTimestepSetRelaxFn step_setrelaxfn;
+  ARKTimestepSetOrder step_setorder;
+  ARKTimestepSetNonlinearSolver step_setnonlinearsolver;
+  ARKTimestepSetLinear step_setlinear;
+  ARKTimestepSetNonlinear step_setnonlinear;
+  ARKTimestepSetNlsRhsFn step_setnlsrhsfn;
+  ARKTimestepSetDeduceImplicitRhs step_setdeduceimplicitrhs;
+  ARKTimestepSetNonlinCRDown step_setnonlincrdown;
+  ARKTimestepSetNonlinRDiv step_setnonlinrdiv;
+  ARKTimestepSetDeltaGammaMax step_setdeltagammamax;
+  ARKTimestepSetLSetupFrequency step_setlsetupfrequency;
+  ARKTimestepSetPredictorMethod step_setpredictormethod;
+  ARKTimestepSetMaxNonlinIters step_setmaxnonliniters;
+  ARKTimestepSetNonlinConvCoef step_setnonlinconvcoef;
+  ARKTimestepSetStagePredictFn step_setstagepredictfn;
+  ARKTimestepGetNumLinSolvSetups step_getnumlinsolvsetups;
+  ARKTimestepGetCurrentGamma step_getcurrentgamma;
+  ARKTimestepGetNonlinearSystemData step_getnonlinearsystemdata;
+  ARKTimestepGetNumNonlinSolvIters step_getnumnonlinsolviters;
+  ARKTimestepGetNumNonlinSolvConvFails step_getnumnonlinsolvconvfails;
+  ARKTimestepGetNonlinSolvStats step_getnonlinsolvstats;
```

Comment:
Relaxation is handled really cleanly now. There's only 1 boolean added to this struct to check for support and all of the getter and setter function are in `relax_mem` to keep this v-table from growing.

For NLS this is handled differently. All the getters and setters are here at the top level memory structure which makes the v-table large, especially for methods that don't need this functionality. I wonder if we can follow a similar approach to relaxation here to have a `nls_mem` with pointers to the functions and to a `SUNNonlinearSolver`.

Reply:
I was thinking of something similar, but I didn't want to wrap too many different things into a single PR.  Since these are part of our internal API, then we can always add/remove them in a follow-on PR without any changes to the user interface.

Technically, all we'd really need is to store the nonlinear solver parameters in a structure that is stored in the main ARKODE memory structure (instead of in the stepper memory).  This could be in a separate struct, like you mentioned, and the main `ARKodeSet` and `ARKodeGet` functions would just set the parameters based on the NULLity of this solver structure (removing need for the vtable functions).

My main concern with moving the nonlinear solver parameters into the main ARKODE memory structure is for your operator-splitting stepper.  What if has multiple implicit sub-steppers, and you want separate nonlinear solver parameters for each?  If ARKODE owned the parameters then there could only exist one set.  Then again, in the current PR then even if the stepper could store multiple copies of the implicit solver parameters, it could only attach a single function to each vtable pointer, allowing `ARKodeSet` control over only one set of parameters.

Reply:
> Since these are part of our internal API, then we can always add/remove them in a follow-on PR without any changes to the user interface.
Yes, good point

> Technically, all we'd really need is to store the nonlinear solver parameters in a structure that is stored in the main ARKODE memory structure (instead of in the stepper memory). This could be in a separate struct, like you mentioned, and the main ARKodeSet and ARKodeGet functions would just set the parameters based on the NULLity of this solver structure (removing need for the vtable functions).
Yes, that's exactly what I'm thinking

> My main concern with moving the nonlinear solver parameters into the main ARKODE memory structure is for your operator-splitting stepper
I don't see any issues with this. The nonlinear solver for each inner solver is stored with the inner solver and not the operator splitting stepper. Similarly, an MRI method does not manage the nonlinear solver for the fast inner solver. The operator splitting stepper will simply disable support for a nonlinear solver. You point would be valid for something like an ADI method, but I don't see any issues with our current steppers.

> Regarding your comment "... which makes the v-table large, especially for methods that don't need this functionality" -- all of these are initialized to NULL, so steppers that don't need these can just ignore them entirely
I'm just pointing out the extra storage for these function pointers (which may all be NULL) which could be eliminated with a `nls_mem`. I think we're on the same page.

Reply:
As you said the `nls_parameters` could be done in a later PR, and I'm not relying on it for the operator splitting. I'll leave that decision to you. I'll finish my initial review today. So far, I really like the new interface structure!

Reply:
Actually, I'll leave this to a subsequent PR, since it will take a bit of focused effort to do this correctly.